### PR TITLE
Add utility for checking bank conflict of shared memory

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Inspired by https://github.com/csarofeen/pytorch/pull/1900, but the approach here is different. The approach here does the checking at compile time, and can be used in unit tests. Sample debug dump:
```
======= Bank confliction =======
Expr: T11_s[( ( ( ( ( i99 * 128 ) + threadIdx.x ) * 2 ) / 32 ) * 32 ), ( ( ( ( i99 * 128 ) + threadIdx.x ) * 2 ) % 32 )] view( T4 )
   = T10_g[( ( ( ( ( blockIdx.x % ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) / 32 ) ) % T0.size[0] ) * ( T0.size[2] * T0.size[1] ) ), ( ( ( ( ( ( ( blockIdx.x / ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) % 32 ) ) / T0.size[2] ) * 16 ) + ( ( ( ( blockIdx.x % ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) / 32 ) ) / T0.size[0] ) ) * T0.size[2] ), ( ( ( ( blockIdx.x / ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) % 32 ) ) % T0.size[2] )] view( T0 );

output conflict: 2 way
Expr: T1_s[( ( ( ( ( i99 * 128 ) + threadIdx.x ) * 2 ) / 32 ) * 32 ), ( ( ( ( i99 * 128 ) + threadIdx.x ) * 2 ) % 32 )] view( T4 )
   = T0_g[( ( ( ( ( blockIdx.x % ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) / 32 ) ) % T0.size[0] ) * ( T0.size[2] * T0.size[1] ) ), ( ( ( ( ( ( ( blockIdx.x / ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) % 32 ) ) / T0.size[2] ) * 16 ) + ( ( ( ( blockIdx.x % ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) / 32 ) ) / T0.size[0] ) ) * T0.size[2] ), ( ( ( ( blockIdx.x / ( ceilDiv(( 16 * T0.size[0] ), 32) ) ) * 32 ) + ( ( ( ( ( i99 + nvfuser_zero ) * 128 ) + threadIdx.x ) * 2 ) % 32 ) ) % T0.size[2] )] view( T0 );

output conflict: 2 way
================================
```

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            uop->in()->as<kir::TensorIndex>(), for_loops_));
+      }
+      if (isSmemTensorIndex(uop->out())) {
+        conflict_ways.second = getConflictWays(evaluateAddressesOnFirstWarp(
+            uop->out()->as<kir::TensorIndex>(), for_loops_));
+      }
+      if (conflict_ways.first > 1 || conflict_ways.second > 1) {
+        bank_conflict_info_[expr] = conflict_ways;
+      }
+    } else if (expr->isA<LoadStoreOp>()) {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp:131]: Not sure if `ld.matrix` is supported correctly here. Will need to dig deeper to understand how it works. I will update this in later PR.

File: torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  for (auto tidx : c10::irange(phase_size)) {
+    int64_t index = 0;
+    ExpressionEvaluator expr_eval(ti->fusion());
+    for (auto fl : for_loops) {
+      if (fl->index()->isA<NamedScalar>() &&
+          fl->index()->as<NamedScalar>()->name() == "threadIdx.x") {
+        expr_eval.bind(fl->index(), tidx);
+      } else {
+        expr_eval.bind(fl->index(), 0);
+      }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp:68]: Does this assume `threadIdx.y/z` don't matter for bank conflicts? What about if we have, e.g., `blockDim.x == 1 && blockDim.y == 32`?

**Comment by naoyam** [comment]
Compile time is nice. Is there any limitation though? Can all bank conflicts be found at the compile time? This doesn't necessarily be perfect, but it'd be important to understand limitations if any. Left a related inline comment.

File: torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp:68]: Yes, I am currently making this assumption. But I can add a new overload that takes a launch parameter, which will lift this assumption.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp:68]: I'm fine with the current state as long as the assumptions are made clear. Not sure how important to make this more flexible at this point.

**Comment by zasdfgbnm** [comment]
> Compile time is nice. Is there any limitation though? Can all bank conflicts be found at the compile time?

Good question!

On the one hand, for our current supported cases, I can not think of any of them that can not be supported at compile time. In the future, if we begin to support crazy things that have data-dependent indexing (such as `T1[T0[i0] + 1]`), then this could become an issue. 

But on the other hand, the bank conflict checking utility as in this PR made many assumptions:
1. `blockDim.x` is large enough to hold one phase
2. The address only depends on loop variables (there can not be a thing like `T0.stride[0]`, `blockDim.x`)
3. The data of the tensor is accessed by `T0[index]`, where `index` is the one stored in the `TensorIndex` object, not sure if there is any case that does not use such an access pattern.
4. Only checking the first iteration, and the start of all loop variables are assumed to be `0` (if we have something like `T1_s[tidx, 5]`, then different iterations should have different results, which this utility will not be able to handle now)
5. All shared memory tensors are allocated starting from a multiple of (4*32)
6. The only source of bank confliction is from within a tensor. There is no bank conflict between different tensors. ~This will fail if we have a lot of tensors like `T1_s[tidx, 2] ca_pos(1)`.~ (Edit: Just checked, `T1_s[tidx, 2] ca_pos(1)` should be handled correctly)

And besides that, this utility is only tested with limited cases (which are the transpose examples I added in unit tests). And if there are cases that could break the above assumption, it is possible that this utility will just generate silent wrong result. So we definitely need to improve this over time, and add tests for more cases. And I would not consider this tool as a single source of trust.

**Comment by zasdfgbnm** [comment]
For the above assumptions, 1 and 2 can be easily lifted by adding new interface to allow user feeding more information, 3 should be improved over time when we add more tests and does more code review. For 4, it would be easy to lift the "start from 0" assumption, and I don't see any value for computing the bank conflict for all iterations (I am not developing a tool to replace nsight compute). ~For 6, does our system support this use case? I can not imagine how the allocation will be done.~(Edit: checked, this is supported well)

File: torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/lower_bank_conflict.cpp:68]: I am also OK with this assumption in this PR. I will write a few followup PRs to lift some of these assumptions.

**Review by naoyam — APPROVED** [review]
Approving as I think this is good enough although it has several assumptions. Please make sure the assumptions are added as comments.

**Pull request merged**